### ISSUE TITLE
Routing with Modules

### ISSUE THREAD

**Issue by NicholasLYang** [SergioBenitez/Rocket#884]
_1. The version of Rocket you're using. Ensure it's the latest, if possible._
v0.4.0
 _2. The operating system (distribution and version) where the issue occurs._
macOS 10.13.6
rustc 1.33.0-nightly (8e2063d02 2019-01-07)

_3. A brief description of the bug:_

I'm trying to put my routes into a separate module. However, when I do so, I get the following error: 
```
error[E0425]: cannot find value `static_rocket_route_info_for_index` in module `routes`
  --> src/main.rs:36:45
   |
36 |         .mount("/projects", routes![routes::index, routes::create])
   |                                             ^^^^^ not found in `routes`
```

I am using the advice in the guide, namely, using a namespaced path.

  _4. How you uncovered the bug. Short, reproducible tests are especially useful._
Basically just working on my project. I can throw together a minimal reproduction if you want.
  _5. Ideas, if any, about what Rocket is doing incorrectly._
Maybe the `get` macro needs to return a public function?

In general, I'm a little confused about how routes should generally be namespaced and organized. I'm coming from a Rails background, so my natural inclination is to organize around resources, i.e. standard CRUD operations on a single model. However, for that, I'd like to organize the difference resource route handlers in proper namespaces. Is this advisable? Or is there a more idiomatic way of organizing routes?

**Comment by jebrosen** [comment]
Is the route function itself declared as public? The exact code that results in that error would definitely help with debugging if you still have trouble.

> However, for that, I'd like to organize the difference resource route handlers in proper namespaces. Is this advisable? Or is there a more idiomatic way of organizing routes?

Rocket doesn't enforce any particular organization. Rocket's examples usually name all routes right in `main`. Another way is to add a `pub fn routes() -> Vec<Route>` function to each module that provides that module's routes.

**Comment by jebrosen** [comment]
I can't seem to get your error message (cannot find value) to show up. I can, however, get ```error[E0603]: static `static_rocket_route_info_for_test` is private``` and the nonsensical ```error[E0423]: expected value, found macro `routes::static_rocket_route_info_for_test` ```.

Would you mind posting `main.rs` as well? Additionally, try after a `cargo update` and the latest nightly, if possible.

**Comment by NicholasLYang** [comment]
Sure! Here's `main.rs`. I tried updating and it didn't fix the problem.

```
#![feature(proc_macro_hygiene, decl_macro, custom_attribute)]
#[macro_use]
extern crate failure_derive;

#[macro_use]
extern crate rocket_contrib;

#[macro_use]
extern crate diesel;

use itertools::Itertools;
use std::fs::File;
use std::io::Read;

mod controllers;
mod github;
mod models;
mod routes;
mod schema;
mod types;

use crate::types::{DbConn, Result, RulesConfig};
use rocket::*;
use rocket_contrib::json::Json;

#[get("/<username>/<repo_name>")]
fn validate_repo(username: String, repo_name: String) -> Result<String> {
    let config = load_config()?;
    let repo = format!("{}/{}", username, repo_name).to_string();
    let issues = github::check_repo(&repo, config.into_rules()?)?;
    Ok(issues.iter().map(ToString::to_string).join("\n"))
}

fn main() {
    rocket::ignite()
        .mount("/projects", routes![routes::index, routes::create])
        .attach(DbConn::fairing())
        .launch();
}

fn read_config_file() -> Result<String> {
    let mut file = File::open("config.toml")?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    Ok(contents)
}

fn load_config() -> Result<RulesConfig> {
    let contents = read_config_file()?;
    let config = toml::from_str(&contents)?;
    Ok(config)
}
```

You can also just check out my project [here](https://github.com/hacknyu/aletheia).

**Comment by SergioBenitez** [comment]
Looks like @jebrosen has identified the issue and proposed a solution, @NicholasLYang. Does that resolve your issue?

**Comment by NicholasLYang** [comment]
Yep! Thank you so much for your help @jebrosen.

**Comment by hajsf** [comment]
I've a question related to `Routing with Modules`, different than the point above, but related to the subject.

Why we are using:
```rust
mod routes;
rocket::ignite().mount("/projects", routes![routes::index, routes::create]);
```

And can not use:
```rust
mod routes::{index, create};
rocket::ignite().mount("/projects", routes![index, create]);
```

Using the last will get the below error:
```
error[E0425]: cannot find value `static_rocket_route_info_for_call` in this scope
  --> src\main.rs:17:29
   |
17 |         .mount("/", routes![index, create]) 
   |                             ^^^^ not found in this scope
```
